## String Boolean Function
   * True or False
   * islower() -- True if the string have all characters lower
   * isupper() -- True if the string has all characters upper
   * istitle() -- True if the string in title case
   * isnumeric() -- True if the strinh is numreic
   * isalpha() -- True if the string only alaphabets
   * isspace() -- True if the string is space

In [1]:
s1 = "python"
print(s1.islower())
print(s1.isupper())

True
False


In [2]:
s1 = "Python Programming"
print(s1.istitle())

True


In [3]:
s1 = "Py123456"
s2 = "Python"
print(s1.isalpha())
print(s2.isalpha())

False
True


In [4]:
s1 = "123456"
s2 = "Python"
print(s1.isnumeric())
print(s2.isnumeric())

True
False


### String Methods
- split() -- returns the list of the string sepearated by whitespace
- replace() -- returns the string and replaces by word/characters
- join() -- concatination of two strings

In [5]:
s1 = "Python Programming Learn Made Easy"
li = s1.split() #default is white space
print(li)

['Python', 'Programming', 'Learn', 'Made', 'Easy']


In [6]:
s1 = "Python Programming Learn Made Easy"
li = s1.split('a')
print(li)

['Python Progr', 'mming Le', 'rn M', 'de E', 'sy']


In [7]:
s1 = "Python Programming Learn Made Easy"
print(s1)
print(s1.replace(' ','-'))

Python Programming Learn Made Easy
Python-Programming-Learn-Made-Easy


In [8]:
s1 = "Python Programming Learn Made Easy"
print(s1)
print(s1.replace('a','@'))

Python Programming Learn Made Easy
Python Progr@mming Le@rn M@de E@sy


In [9]:
s1 = "python"
print(" ".join(s1))

p y t h o n


### String Formatting
- Classic version
- Current Version Python

In [10]:
 # CLassic Version (C or C++)
li = ["Python","Programming"]
print('%s %s' % (li[0],li[1]))


Python Programming


In [11]:
li = [1,2,3,4]
print('%d %d %d %d' % (li[0],li[1],li[2],li[3]))

1 2 3 4


In [12]:
# .format
li = ["Python","Programming"]
print("List Item 0 : {0} \nList Item 1 : {1}".format(li[0],li[1]))

List Item 0 : Python 
List Item 1 : Programming


In [13]:
li = [1,2,3,4]
print('{0} {1} {2} {3}'.format(li[0],li[1],li[2],li[3]))

1 2 3 4


### Data structeres
- List
- Tupples
- Dictionaries

### Dictionaires 
- It works on the concept of unique data
- Key-Value Pair
- Eact Key is separated by Colon(:)
- keach key and value is separated by comma(,)
- Dictionaries are enclosed with Curly brackets({})

In [14]:
d1 = {'Name' : "Madhu", "EmailID" : "Mads@gimail.com", "Address" : "Vizag"}
print(d1)

{'Name': 'Madhu', 'EmailID': 'Mads@gimail.com', 'Address': 'Vizag'}


In [15]:
d1["Name"] #Access the specific key

'Madhu'

In [16]:
d1["EmailID"] = "Madhu@gmail.com" #Update the value using key
print(d1)

{'Name': 'Madhu', 'EmailID': 'Madhu@gmail.com', 'Address': 'Vizag'}


In [17]:
li = d1.keys() #All the keys of dict. object
print(li)

dict_keys(['Name', 'EmailID', 'Address'])


### Contact Application
- Add Contact Details
- search of contact deails
- modify
- delete

In [18]:
contacts = {} # Empty Dict Object
# Lets Create A function To add details Details
def addContact(name,phone):
    if name not in contacts: #contion to check if name exists
        contacts[name] = phone
        print('contact details are added successfully')
    else:
        print('name already exists')
    return
addContact('Madhu','1234567890')
addContact('Swathi','0987654321')
addContact('Madhu','7890654311')

contact details are added successfully
contact details are added successfully
name already exists


In [19]:
def searchContact(name):
    if name in contacts:
        print(name," : ",contacts[name])
    else:
        print('{0} does not exits' .format(name))
        #print('%s does not exits' % contacts([name]))
    return
searchContact('Madhu')
searchContact('gitam')

Madhu  :  1234567890
gitam does not exits


In [27]:
# Import new contact details
# Merge the previous details with new details
def importContacts(newContacts):
    contacts.update(newContact)
    print(len(newContacts.keys()),'Contacts are added')
    return
newContacts = {'Gitam':6789054321,'Anil':5678904321,'Python':123097685}
importContacts(newContacts)

3 Contacts are added


In [28]:
contacts

{'Madhu': '1234567890',
 'Swathi': '0987654321',
 'Gitam': 6789054321,
 'Anil': 5678904321,
 'Python': 123097685}

In [30]:
def modifyContact(name,phone):
   if name in contacts:
       contacts[name] = phone
       print(name,'updated successfully')
   else:
       print(name,'not exists')
   return
modifyContact('Madhu',9876543210)
modifyContact('Navven',6541236540)

Madhu updated successfully
Navven not exists


In [34]:
def deleteContact(name):
    if name in contacts:
        del contacts[name]
        print(name,'Deleted Successfully')
    else:
        print(name,'Not Present')
    return
deleteContact('Madhu')

Madhu Deleted Successfully


In [32]:
contacts

{'Madhu': 9876543210,
 'Swathi': '0987654321',
 'Gitam': 6789054321,
 'Anil': 5678904321,
 'Python': 123097685}

In [35]:
contacts

{'Swathi': '0987654321',
 'Gitam': 6789054321,
 'Anil': 5678904321,
 'Python': 123097685}

## Tupples
 - Tuples encolsed with parenthesis
### Dffrence between Tuples and List
- list are mutable -- can be changed / modified
- e adnd Write the data
- Tuples are immutable -- cant be changed / modififd
 - read data only

In [36]:
t1 = (1,2,3,4,5)
print(t1)

(1, 2, 3, 4, 5)


In [37]:
print(t1[1:-1])

(2, 3, 4)


### Standard libraries
- Regular Expressions
 - To validate the indian Mobile
 - To validate the Email id
 - To validate the Username
 - To validate the  password
- Understanding the Regular Expression
 

In [39]:
# Function to test the two digit number as input
# True -- Input two digit number
# False -- Input not a two digit number
import re
def twoDigitMatching(n):
    pattern = '^[0-9]{2}$'
    n = str(n)
    if re.match(pattern,n):
        return True
    return False
print(twoDigitMatching(12)) #TRue
print(twoDigitMatching(123))#False

True
False


In [40]:
#Regular Expression for username
# Username -- Should contains lower and upper case characters
# Length -- min 5 max 12
def validateUsername(s):
   pattern = '^[a-zA-Z]{5-12}$'
   if re.match(pattern,s):
       return True
   return False
print(validateUsername('AnilKumar')) # True
print(validateUsername('Anil'))     # False
print(validateUsername('GitamVizagPythonProgramming')) # False

False
False
False


### Regular Expression for Indian Mobile Number
- 10 Digit Number
    - First Number starts from [6-9]
    - Rest of 9 numbers is between [0-9]
        - exapmle : 987654321 #True
        - Example : 570375531 #False
- 11 Digit Number
    - First number can be 0
    - Second number can be [6-9]
    - Rest of 9 numbers is between [0-9]
        - exapmle : 0987654321 #True
        - Example : 570375531 #False
- 13 Digit Number
    - First number can be +
    - Second number 9
    - Third number 1
    - Fourth number[6-9]
    - Rest numbe [0-9]


In [ ]:
# Rule 1 : ^[6-9][0-9]{9}$
# Rule 2 : ^[0][6-9][0-9]{9}$
# Rule 3 : ^[+][9][6-9][0-9]{9}$

In [44]:
import re
def validateIndianPhone(n):
    pattern = '^[6-9][0-9]{9}$|^[0][6-9][0-9]{9}$|^[+][9][1][6-9][0-9]{9}$'
    n = str(n)
    if re.match(pattern,n):
        return True
    return False
validateIndianPhone('+919988774455')

True

### Validate the EmailId
- Example : Username@DomianName.extension
- Example : dsps@gmail.com
- username :
    - Len of string will be [6-15]
    - No special characters except Underscore(_) and dot(.)
    - Username should not begin with Underscore(_)
    - All characters set Digits,Lower case and Uppercase
        - Example : Mads_100
- DominName
    - Len of string will be [3-18]
    - No spls characters
    - Characters set will be Digits,Lower case and Uppercase
        - Example : gmail,hotmail.ymail ect
- Extensions :
    - Len of string will be[2-4]
    - No spls characters
    - Characters set will be Digits,Lower case and Uppercase
        -Example : .com,.org , .in etc
    

In [ ]:
# Username RE :  [0-9a-zA-Z][0-9a-zA_.]{5,14}
# Domain Name RE : [a-z0-9]{3,18}
# Extension RE : [a-z]{2,4}
# EmailId RE : ^[0-9a-zA-Z][0-9a-zA_.]{5,14}[@][a-z0-9]{3,18}[.][a-z]{2,4}$

In [46]:
def validateEmail(email):
    pattern = '^[0-9a-zA-Z][0-9a-zA_.]{5,14}[@][a-z0-9]{3,18}[.][a-z]{2,4}$'
    if re.match(pattern,email):
        return True
    return False
validateEmail('anilkumar1889@hotmail.com')

True

### Python Graphics
- lib - turle

### Package , Module and Sub Package
#### Package
    - A Collection of Modules(Single Python file .py)
#### Module
    - A Single Python File contains group function
    - Package -> Sub Package -> Modules -> Functions -> Statements
    

In [47]:
import math
math.floor(123.456)

123

In [48]:
math.ceil(123.456)

124

In [50]:
from math import factorial as fact
fact(10)

3628800

In [51]:
from math import gcd as gcd
gcd(10,15)

5

In [4]:
# import random package
import random
# Function to generate  the N number in given range
def generateRandNum(n,lb,ub):
   for i in range(0,n):
       print(random.randint(lb,ub),end=' ')
   return
generateRandNum(10,0,100)

51 46 45 61 34 47 89 38 14 25 

In [ ]:
# Creat a simple game
# Try to generate 20 numbers
# Input : Number
# Present : Congrats!!!!
# Not present : Try Again   

In [6]:
from random import randint as rdint
def generateNumber(n,lb,ub):
    li = []
    for i in range(0,n):
        li.append(rdint(lb,ub))
    return li
def check(n):
    li = generateNumber(20,100,999)
    if n in li:
        print("Congrats!!!")
    else:
        print("Try Again!!!")
    return
check(569) 

Try Again!!!


In [7]:
import turtle as t
a1 = t.Turtle()
for i in range(40):
   a1.forward(50)
   a1.right(144)
t.done()

Terminator: 